File: js/src/modules/analyzer.js
/**
 * Statistical analysis module for VNG data - JavaScript version
 * Compares metrics across multiple files and calculates statistics
 */

const { calculateStdDev, calculatePercentChange } = typeof module !== 'undefined' && module.exports
    ? require('../utils/statistics')
    : window.VNGStatistics;

/**
 * Analyzes the data from multiple file maps.
 * @param {Array} fileDataList - List of dictionaries, each containing parsed file data
 *                             Format: [{'name': string, 'data': {category: {metric: {value, isFlagged}}}}]
 * @returns {Object} Nested dictionary with analysis results:
 *                   {category: {metric: {values: [], flags: [], delta: number|null,
 *                              stdDev: number|null, percentChange: number|null}}}
 */
function runAnalysis(fileDataList) {
    const results = {};

    if (!Array.isArray(fileDataList) || fileDataList.length === 0) {
        return results;
    }

    const firstMap = fileDataList[0]['data'];

    // 1. Find common category:metric pairs ("Apples to Apples")
    // Build a set of all (category, metric) pairs from the first file
    const commonPairs = new Set();

    for (const category in firstMap) {
        for (const metric in firstMap[category]) {
            commonPairs.add(`${category}::${metric}`);
        }
    }

    // Check which pairs exist in all files
    for (let i = 1; i < fileDataList.length; i++) {
        const fileData = fileDataList[i];
        const filePairs = new Set();

        for (const category in fileData['data']) {
            for (const metric in fileData['data'][category]) {
                filePairs.add(`${category}::${metric}`);
            }
        }

        // Keep only pairs that exist in this file too
        const newCommonPairs = new Set();
        for (const pair of commonPairs) {
            if (filePairs.has(pair)) {
                newCommonPairs.add(pair);
            }
        }
        commonPairs.clear();
        for (const pair of newCommonPairs) {
            commonPairs.add(pair);
        }
    }

    // 2. Populate results and calculate stats
    for (const pair of commonPairs) {
        const [category, metric] = pair.split('::');

        // Extract values and flags separately
        const values = [];
        const flags = [];

        for (const fileData of fileDataList) {
            // Check if category and metric exist in this file
            if (fileData['data'] && fileData['data'][category] && fileData['data'][category][metric]) {
                const metricData = fileData['data'][category][metric];
                values.push(metricData['value']);
                flags.push(metricData['isFlagged']);
            } else {
                // Skip this file if it doesn't have this metric
                continue;
            }
        }

        // Only process if we have data from all files
        if (values.length !== fileDataList.length) {
            continue;
        }

        // Calculate statistics
        let delta = null;
        let stdDev = null;
        let percentChange = null;

        if (fileDataList.length === 2) {
            delta = values[1] - values[0];
            percentChange = calculatePercentChange(values[0], values[1]);
        }

        if (fileDataList.length >= 2) {
            stdDev = calculateStdDev(values);
        }

        // Ensure category exists in results
        if (!(category in results)) {
            results[category] = {};
        }

        // Store metric data
        results[category][metric] = {
            'values': values,
            'flags': flags,
            'delta': delta,
            'stdDev': stdDev,
            'percentChange': percentChange
        };
    }

    return results;
}

// Export function
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { runAnalysis };
} else if (typeof window !== 'undefined') {
    window.VNGAnalyzer = { runAnalysis };
}

File: _python_backup/modules/analyzer.py
"""
Statistical analysis module for VNG data
Compares metrics across multiple files and calculates statistics
"""

from typing import Dict, List, Optional, Any
from utils.statistics import calculate_std_dev, calculate_percent_change


def run_analysis(file_data_list: List[Dict[str, Dict[str, Dict[str, Any]]]]) -> Dict[str, Dict[str, Dict[str, Any]]]:
    """
    Analyzes the data from multiple file maps.
    
    Args:
        file_data_list: List of dictionaries, each containing parsed file data
                       Format: [{'name': str, 'data': {category: {metric: {value, is_flagged}}}}]
    
    Returns:
        Nested dictionary with analysis results:
        {category: {metric: {values: [], flags: [], delta: float|None, 
                             std_dev: float|None, percent_change: float|None}}}
    """
    results: Dict[str, Dict[str, Dict[str, Any]]] = {}
    
    if not file_data_list or len(file_data_list) == 0:
        return results
    
    first_map = file_data_list[0]['data']
    
    # 1. Find common category:metric pairs ("Apples to Apples")
    # Build a set of all (category, metric) pairs from the first file
    common_pairs = set()
    for category, metrics in first_map.items():
        for metric in metrics.keys():
            common_pairs.add((category, metric))
    
    # Check which pairs exist in all files
    for file_data in file_data_list[1:]:
        file_pairs = set()
        for category, metrics in file_data['data'].items():
            for metric in metrics.keys():
                file_pairs.add((category, metric))
        # Keep only pairs that exist in this file too
        common_pairs = common_pairs.intersection(file_pairs)
    
    # 2. Populate results and calculate stats
    for category, metric in common_pairs:
        # Extract values and flags separately
        values = []
        flags = []
        
        for file_data in file_data_list:
            # Check if category and metric exist in this file
            if category in file_data['data'] and metric in file_data['data'][category]:
                metric_data = file_data['data'][category][metric]
                values.append(metric_data['value'])
                flags.append(metric_data['is_flagged'])
            else:
                # Skip this file if it doesn't have this metric
                continue
        
        # Only process if we have data from all files
        if len(values) != len(file_data_list):
            continue
        
        # Calculate statistics
        delta = None
        std_dev = None
        percent_change = None
        
        if len(file_data_list) == 2:
            delta = values[1] - values[0]
            percent_change = calculate_percent_change(values[0], values[1])
        
        if len(file_data_list) >= 2:
            std_dev = calculate_std_dev(values)
        
        # Ensure category exists in results
        if category not in results:
            results[category] = {}
        
        # Store metric data
        results[category][metric] = {
            'values': values,
            'flags': flags,
            'delta': delta,
            'std_dev': std_dev,
            'percent_change': percent_change
        }
    
    return results

